fairscale/experimental/nn/sync_batchnorm.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Any, Dict, Optional, Tuple
+
+import torch
+import torch.distributed as dist
+from torch.distributed import ProcessGroup
+
+if torch.__version__.split(".")[:2] >= ["1", "8"]:
+    from torch.distributed import all_reduce
+else:
+    # Copied from https://github.com/pytorch/pytorch/blob/v1.8.1/torch/distributed/nn/functional.py
+    class _AllReduce(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, op, group, tensor):  # type: ignore
+            ctx.group = group
+            ctx.op = op
+            tensor = tensor.clone()
+            dist.all_reduce(tensor, op=op, group=group)
+            return tensor
+
+        @staticmethod
+        def backward(ctx, grad_output):  # type: ignore
+            return (None, None) + (_AllReduce.apply(ctx.op, ctx.group, grad_output),)
+
+    def all_reduce(tensor, op=dist.ReduceOp.SUM, group=dist.group.WORLD):  # type: ignore
+        return _AllReduce.apply(op, group, tensor)
+
+
+def _forward(
+    input: torch.Tensor,
+    affine: bool,
+    track_running_stats: bool,
+    mean: torch.Tensor,
+    var: torch.Tensor,
+    momentum: float,
+    eps: float,
+    weight_param: torch.Tensor,
+    bias_param: torch.Tensor,
+    running_mean: torch.Tensor,
+    running_var: torch.Tensor,
+    total_count: torch.Tensor,
+) -> torch.Tensor:
+    scale = torch.rsqrt(var + eps)
+    bias = -mean * scale
+    if affine:
+        scale = scale * weight_param.reshape(mean.shape)
+        bias = bias + bias_param.reshape(mean.shape)
+
+    if track_running_stats:
+        with torch.no_grad():
+            unbiased_var = var * (total_count / (total_count - 1))
+            running_mean += momentum * (mean.reshape(-1) - running_mean)
+            running_var += momentum * (unbiased_var.reshape(-1) - running_var)
+
+    return input * scale + bias
+
+
+if torch.__version__.split(".")[:2] >= ["1", "7"]:
+    _forward = torch.jit.script(_forward)  # type: ignore
+
+
+class SyncBatchNorm(torch.nn.BatchNorm2d):
```

Comment:
any reason we decided to subclass BatchNorm2d? Why not subclass torch.nn.SyncBatchNorm so that we could use the `convert_sync_batchnorm `. Are there things that we have modified in this function?

Reply:
Unfortunately SyncBatchNorm's `convert_sync_batchnorm` explicitly calls `SyncBatchNorm` instead of `cls` so sub-classing it would still require implementing `convert_sync_batchnorm`. I will modify to sub-class `_BatchNorm` although I'm hesitant to sub-class a private class. I didn't want to sub-class SyncBatchNorm because it felt odd to implement `SyncBatchNorm` which sub-classes `SyncBatchNorm`. I was also worried about future changes to `SyncBatchNorm` breaking my sub-class. Felt safer to sub-class on of the other classes. Sub-classes `_BatchNorm` might be the least-worst option.